README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# mipred

[![Build
Status](https://travis-ci.org/BARTJAMertens/mipred.svg?branch=master)](https://travis-ci.org/BARTJAMertens/mipred)

<!-- Check https://www.r-pkg.org/services   for badges -->

<!-- OR:    https://cran.r-project.org/web/packages/badgecreatr/vignettes/all_badges.html -->

<!-- [![CRAN_Status_Badge](https://www.r-pkg.org/badges/version-last-release/mipred)](https://CRAN.R-project.org/package=mipred) -->

<!-- [![](https://cranlogs.r-pkg.org/badges/grand-total/mipred)](https://CRAN.R-project.org/package=mipred) -->

<!-- [![Download counter](http://cranlogs.r-pkg.org/badges/mipred)](https://cran.r-project.org/package=mipred) -->

<!-- [![Rdoc](http://www.rdocumentation.org/badges/version/mipred)](http://www.rdocumentation.org/packages/mipred) -->

The goal of mipred is to calibrate a prediction rule using generalized
linear models or Cox regression modeling, using multiple imputation to
account for missing values in the predictors as described by Mertens,
Banzato and de Wreede (2018) (<https://arxiv.org/abs/1810.05099>).
Imputations are generated using the R package mice without using
outcomes on observations for which the prediction is generated. Two
options are provided to generate predictions. The first is
prediction-averaging of predictions calibrated from single models fitted
on single imputed datasets within a set of multiple imputations. The
second is application of the Rubin’s rules pooled model. For both
implementations, unobserved values in the predictor data of new
observations for which the predictions are derived are automatically
imputed. The package contains two basic workhorse functions, the first
is ‘mipred’ which generates predictions of outcome on new observations
(when outcomes will by definition usually not be available at the time
of calibration of the prediction rule). The second is the function
‘mipred.cv’ which generates cross-validated predictions with the
methodology on existing data for which outcomes have already been
observed. This allows users to assess predictive potential of the
prediction models which are calibrated. The present version of the
package is preliminary (development) and has only been thoroughly
checked for application on binary-outcome logistic regression for now.
The vignette which is included documents application of the functions
for binary outcome data. Although we did not check extensively, the
package should also work for continuous and counting outcomes. We are
working to expand the functionality to censored survival
outcomes.

## Installation

<!-- You can install the released version of mipred from [CRAN](https://CRAN.R-project.org) with: -->

You can install the current version into R from GitHub using devtools:

``` r
library(devtools)
devtools::install.github("BartJAMertens/mipred")
```

You may need to install and load the devtools package first before using
the above command. See the book “R packages” (online version) by Hadley
Wickham, chapter “Git and Github”.

## Main functions

There are currently two key functions

mipred() \# prediction calibration with multiple imputation for missing
predictors  
mipred.cv() \# cross-validation for prediction calibration with multiple
imputation for missing predictors

The first function calibrates predictions for new observations and
accounts for missing values in the predictor data (of either the
calibration or new validation sample) through multiple imputation. The
second function implements cross-validation of the same approach.

## Example

Let `dataset` be a data.frame consisting of a vector of binary outcomes
`outcome` and two predictors `x1` and `x2`. Likewise, let `newdataset`
be a data.frame with new observations for which the same predictors `x1`
and `x2` are observed and for which we want to predict outcome, using a
model fitted to the old data in `dataset`.

We can generate predictions using the
    command

    preds <- mipred(outcome ~ x1 + x2, family=binomial, data=dataset, newdata=newdataset, nimp=100)

This will use the logistic regression model and 100 imputations.

If we wanted to generate cross-validated predictions within the set
`dataset`, then we can generate these with the same model
    using

    preds.cv <- mipred.cv(outcome ~ x1 + x2, family=binomial, data=dataset,  nimp=100,  folds=10)

This will generate cross-validated predictions from the same model and
100 imputataions for each predicted observation, using 10-folds.

Please refer to the example included with the package. The package also
includes a vignette which documents use for binary outcome
data.

<!-- This is a basic example which shows you how to solve a common problem: -->

<!-- ```{r example} -->

<!-- ## basic example code -->

<!-- ``` -->

<!-- What is special about using `README.Rmd` instead of just `README.md`? You can include R chunks like so: -->

<!-- ```{r cars} -->

<!-- summary(cars) -->

<!-- ``` -->

<!-- You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date. -->

<!-- You can also embed plots, for example: -->

<!-- ```{r pressure, echo = FALSE} -->

<!-- plot(pressure) -->

<!-- ``` -->

<!-- In that case, don't forget to commit and push the resulting figure files, so they display on GitHub! -->


NEWS.md
NEWS
================
Bart JA Mertens
6/29/2019

# mipred 0.0.1.0001

First release of the mipred package. Present version only supports
analysis of continuous, binary and counting outcomes.


tests/testthat.R
library(testthat)
library(mipred)

test_check("mipred")


tests/testthat/test-seeduse-mipred.R
context("use of seed")

skip('seeduse tests mipred')

cll <- readRDS("CLL_BMJ.rds")

cll_bin<-cll

cll_bin$srv5y_s[cll_bin$srv5y>12] <- 0  # Apply an administrative censorship at t=84 months
cll_bin$srv5y[cll_bin$srv5y>12]  <- 12

cll_bin$Status[cll_bin$srv5y_s==1]<- 1  # Define the "Status" variable
cll_bin$Status[cll_bin$srv5y_s==0] <- 0  # Ss numeric -> 1:Dead, 0:Alive

cll_bin$Censor <- NULL
cll_bin$srv5y <- NULL
cll_bin$srv5y_s <- NULL

seeds <- c(7350, 5880, 2571, 3887, 9745, 5608, 3468, 2711, 3448, 4676,
  1627, 5087, 6479, 3901, 1037, 6622, 4832, 3382, 8515, 268, 3210,
  5944, 3380, 3646, 9188, 2750, 7987, 466, 2981, 136, 5134, 8476,
  5618, 8948, 1590, 5281, 2551, 2052, 1444, 8688, 1691, 6398, 4728,
  8267, 5274, 4154, 982, 7540, 3081, 5286, 6405, 1238, 5589, 8221,
  6026, 1922, 7651, 3100, 4341, 2134, 9685, 4524, 9750, 7618, 8211,
  2438, 824, 6982, 6548, 9743, 4302, 9038, 185, 4418, 1306, 6643,
  9865, 3436, 6151, 9647, 4261, 8109, 7573, 127, 571, 4545, 7245,
  4764, 5038, 9037, 6085, 1459, 5124, 9463, 942, 2241, 6492, 3190,
  8229, 7099)

output <- list(pred = structure(c(0.272822624187677, 0.26335349393881,
  0.245055602778557, 0.316090037563377, 0.313007499662245, 0.297912651367598,
  0.307891918762763, 0.255163876316593, 0.331122267440826, 0.364964866709756
), .Dim = c(5L, 2L), .Dimnames = list(c("501", "502", "503",
  "504", "505"), NULL)), linpred = structure(c(-0.980348575240286,
    -1.02861092500769, -1.1251589655607, -0.771799173093083, -0.786096224788477,
    -0.857257497214206, -0.809993321899551, -1.07125823500496, -0.703113586720463,
    -0.553879232018371), .Dim = c(5L, 2L), .Dimnames = list(c("501",
      "502", "503", "504", "505"), NULL)))

mice.options <- NULL
mice.options$seed <- as.vector(matrix(seeds,ncol=1,nrow=5*5))[1:4]
mice.options$maxit <- 2

set.seed(12345)
test_that("output replicates from specified seed for mipred and averaging method",{
  expect_equal(mipred(Status~age10+cyto,family=binomial,data=cll_bin[1:500,-1],newdata=cll_bin[501:505,c(-1,-10)], nimp=2, folds=2, mice.options=mice.options)[2:3],output)
    })

output <- list(pred = structure(c(0.313619237134538, 0.287549302953276,
  0.243573611866073, 0.313856654363961, 0.278459046089519, 0.313619237134538,
  0.287549302953276, 0.243573611866073, 0.313856654363961, 0.324948659032894
), .Dim = c(5L, 2L), .Dimnames = list(c("501", "502", "503",
  "504", "505"), NULL)), linpred = structure(c(-0.783252892154253,
    -0.907316378553107, -1.13318601836123, -0.782150197640975, -0.952118141968731,
    -0.783252892154253, -0.907316378553107, -1.13318601836123, -0.782150197640975,
    -0.731121551027735), .Dim = c(5L, 2L), .Dimnames = list(c("501",
      "502", "503", "504", "505"), NULL)))

mice.options <- NULL
mice.options$seed <- as.vector(matrix(seeds,ncol=1,nrow=5*5))[1:2]
mice.options$maxit <- 2

set.seed(12345)
test_that("output replicates from specified seed for mipred and rubin method",{
  expect_equal(mipred(Status~age10+cyto,family=binomial,data=cll_bin[1:500,-1],newdata=cll_bin[501:505,c(-1,-10)], nimp=2, folds=2, mice.options=mice.options, method="rubin")[2:3],output)
})


R/data.R
#' CLL data
#'
#' A dataset containing survival outcome and predictors on 694 patients who
#' received hematopoietic stem cell transplant.
#'
#' @format A data frame with 694 rows and 11 variables. Each row describes the data from a single patient. The below described variables are included in the data file. Missing observations are present in the variables performance status(9\%), remission status (6\%) and cytogenic abnormality (25\%).
#' \describe{
#'   \item{id}{record identification number}
#'   \item{age10}{age at transplantation}
#'   \item{perfstat}{performance status indicated by the Karnofsky Index (four categories)}
#'   \item{remstat}{remission status at transplantation (three categories)}
#'   \item{cyto}{cytogenetic abnormalities (four categories)}
#'   \item{asct}{previous autologous transplantation (two categories)}
#'   \item{donor}{donor type (three categories)}
#'   \item{sex_match}{patient-donor sex match (four categories)}
#'   \item{cond}{conditioning regimen (three categories)}
#'   \item{srv5y}{overall survival (OS) up to five years after first allogeneic stem cell transplantation}
#'   \item{srv5y_s}{censoring indicator (0=alive at end follow-up, 1=dead)}
#' }
#' @references Please reference the following papers when using this data. Schetelig, J. et al. (2017) Risk factors for treatment failure after allogeneic transplantation of patients with CLL: a report from the European Society for Blood and Marrow Transplantation. {Bone Marrow Transplantation},  52, 552-560. Schetelig, J. et al. (2017) Centre characteristics and procedure-related factors have an impact on outcomes of allogeneic transplantation for patients with CLL: a retrospective analysis from the European Society for Blood and Marrow Transplantation (EBMT). {British Journal of Haematology},  178, 521-533. Mertens, B.J.A. et al. (2019) Construction and assessment of prediction rules for binary outcome in the presence of missing predictor data using multiple imputation and cross-validation: theoretical perspective and data-based evaluation. {Biometrical Journal}. See ArXiv for an early version \url{https://arxiv.org/abs/1810.05099}.  We thank EBMT and DKMS for their work in collecting and preparing the CLL data and for approval to share the data.
#' @source European Society for Blood and Marrow Transplantation (EBMT). \url{https://www.ebmt.org}
#'
#'
#' @docType data
"cll"


R/mipred.cv.R
#' Cross-validation prediction using multiple imputation
#'
#' Calculates cross-validated predictions based on within-sample assessment and calibration using
#' generalized linear models with multiple imputations to account for missing values
#' in predictor data.
#'
#' @param formula A formula object providing a symbolic description of the
#'   prediction model to be fitted.
#' @param family Specification of an appropriate error distribution and link
#'   function.
#' @param data A data.frame containing calibration data on \code{n} samples.
#'   Variables declared in \code{formula} must be found in \code{data}.
#' @param nimp Number of imputations used in the prediction of each observation.
#' @param folds Number of fold-partitions defined within \code{data} used in cross-validation.
#' An integer from 2 to \code{n}. Defaults to NULL which internally
#' sets \code{folds=n}, which puts each observation in \code{data}
#' into its own singleton fold for leave-one-out cross-validation.
#' @param method Imputation combination method. This defaults to
#'   \code{"averaging"} for the prediction-averaging approach. The alternative
#'   \code{"rubin"} applies the Rubin's rules pooled model.
#' @param mice.options Optional list containing arguments to be supplied to \code{mice}. Refer to the \code{mice} documentation for details.
#' The following options may be specified: \code{method}, \code{predictorMatrix}, \code{blocks},
#' \code{visitSequence}, \code{formulas}, \code{blots}, \code{post}, \code{defaultMethod},
#' \code{maxit}, \code{printFlag}, \code{seed}, \code{data.init}. Please refer to the
#' \code{mice} documentation for the description of these options. To set the number
#' of imputations \code{nimp} should be used. \code{seed} may be specified as a numeric vector
#' of length \code{nimp*folds} when \code{method} is set to \code{averaging} and of length \code{folds}
#' when \code{method} is set to \code{rubin}. Setting \code{seed} to a vector will cause each next
#' call to \code{mice} to use the next seed value in the vector. Setting the seed to a single
#' numeric value will cause all instances of
#' mice to use that same seed value. If you specify a seed vector of insufficient length
#' then the values will be recycled. The required length is \code{folds*nimp} for the averaging
#' approach and length \code{folds} for the rubin approach. The \code{defaultMethod} is set to
#' \code{c("pmm", "logreg", "polyreg", "polr")} by default. The default setting for
#' \code{printFlag} is FALSE. The default for \code{maxit} is 50. All other options are set
#' to \code{NULL} by default.
#'
#' @return A list containing predictions.
#' \describe{\item{\code{pred}}{Matrix of predictions on the scale of the response
#' variable of dimension \code{n} by \code{nimp}.}
#' \item{\code{linpred}}{Matrix of predictions on the scale
#'   of the linear predictor of dimension \code{n} by \code{nimp}.}}
#'
#' @author Bart J A Mertens, \email{b.mertens@lumc.nl}
#' @references \url{https://arxiv.org/abs/1810.05099}
#' @seealso \code{\link{mice}}
#'
#' @examples
#' \donttest{
#' # Generate a copy of the cll data and construct binary outcome from survival information
#' cll_bin<-cll
#' cll_bin$srv5y_s[cll_bin$srv5y>12] <- 0  # Apply administrative censorship at t=12 months
#' cll_bin$srv5y[cll_bin$srv5y>12]  <- 12
#' cll_bin$Status[cll_bin$srv5y_s==1]<- 1  # Define the new binary "Status" outcome variable
#' cll_bin$Status[cll_bin$srv5y_s==0] <- 0  # As numeric -> 1:Dead, 0:Alive
#' cll_bin$Censor <- NULL # Remove survival outcomes
#' cll_bin$srv5y <- NULL
#' cll_bin$srv5y_s <- NULL
#'
#' # Cross-validate prediction using logistic regression in the first 100 samples
#' # Apply prediction-averaging using 5 imputations, 5 folds and maxit=5.
#' # Note these settings are only for illustration and should be set to higher values for
#' # practical use, particularly for nimp.
#' output<-mipred.cv(Status ~ age10+cyto, family=binomial, data=cll_bin[1:100,-1],
#' nimp=5, folds=5, mice.options=list(maxit=5))
#' }
#'
#' @export
#' @importFrom mice mice complete
#' @importFrom stats glm model.matrix as.formula predict.glm
#'
mipred.cv <-
  function(formula,
    family,
    data,
    nimp,
    folds = NULL,
    method = "averaging",
    mice.options = NULL) {
    call <- match.call()

    if (missing(formula))
      stop("formula not provided")
    if (!inherits(formula, "formula"))
      stop("formula argument not recognized as formula")

    if (is.character(family))
      family <-
        get(family, mode = "function", envir = parent.frame())
    if (is.function(family))
      family <- family()
    if (is.null(family$family)) {
      print(family)
      stop("family not recognized")
    }
    if (missing(data))
      stop("data not provided")
    if (!is.data.frame(data))
      stop("data should be a data frame", call. = FALSE)
    dup <- duplicated(colnames(data))
    if (any(dup))
      stop("Duplicate names found in data: ",
        paste(colnames(data)[dup], collapse = ", "))
    rownames.data <- rownames(data)
    if (missing(nimp))
      stop("nimp not provided")
    if (!is.numeric(nimp) | !is.vector(nimp)) {
      stop("nimp must be integer number greater than zero")
    } else {
      if (!(length(nimp) == 1)) {
        stop("nimp must be integer number greater than zero")
      } else {
        if ((as.integer(nimp) != nimp) | (nimp < 1))
          stop("nimp must be integer number greater than zero")
      }
    }

    if (is.null(folds)) {
      folds <-
        nrow(data) # force default to leave-one-out when NULL
    } else {
      if (!is.numeric(folds) | !is.vector(folds)) {
        stop("folds must be integer number greater than 1")
      } else {
        if (!(length(folds) == 1)) {
          stop("folds must be integer number greater than 1")
        } else{
          if ((as.integer(folds) != folds) | (folds < 2))
            stop("folds must be integer number greater than 1")
        }
      }
    }
    if (folds > nrow(data))
      stop("folds too large")

    if (is.null(mice.options)) {
      # setting default mice options
      mice.options[["method"]] <- NULL
      mice.options[["where"]] <- NULL
      mice.options[["visitSequence"]] <- NULL
      mice.options[["blots"]] <- NULL
      mice.options[["post"]] <- NULL
      mice.options[["defaultMethod"]] <-
        c("pmm", "logreg", "polyreg", "polr")
      mice.options[["maxit"]] <- 50
      mice.options[["printFlag"]] <- FALSE
      mice.options[["seed"]] <- NA
      mice.options[["data.init"]] <- NULL
    } else {
      # updating user provided mice options
      if (is.null(mice.options[["defaultMethod"]]))
        mice.options$defaultMethod <-
          c("pmm", "logreg", "polyreg", "polr")
      if (is.null(mice.options[["maxit"]]))
        mice.options$maxit <- 50
      if (is.null(mice.options[["printFlag"]]))
        mice.options$printFlag <- FALSE
      if (is.null(mice.options[["seed"]]))
        mice.options$seed <- NA
    }
    if (!is.null(mice.options[["m"]]))
      warning("mice.options argument m ignored, using nimp instead")

    if (all(!is.na(mice.options[["seed"]]))) {
      if (!is.vector(mice.options[["seed"]])) {
        stop("in mice.options seed must be a vector")
      } else {
        if (!is.numeric(mice.options[["seed"]])) {
          stop("in mice.options seed must be a numeric vector")
        } else {
          if (!all(as.integer(mice.options[["seed"]]) == mice.options[["seed"]]) |
              any(mice.options[["seed"]] < 0)) {
            stop("in mice.options seed must be positive integer")
          } else {
            if (method == "rubin") {
              mice.options[["seed"]] <-
                matrix(mice.options[["seed"]],
                  nrow = folds,
                  ncol = 1)[, 1]
            } else {
              mice.options[["seed"]] <-
                matrix(mice.options[["seed"]],
                  nrow = folds * nimp,
                  ncol = 1)[, 1]
            }
          }
        }
      }
    }

    # generate calibration data.frame
    mf <- match.call(expand.dots = FALSE)
    m <-
      match(c("formula", "data", "subset", "weights"), names(mf), 0L)
    mf <- mf[c(1L, m)]
    mf$na.action = quote(na.pass)
    mf$drop.unused.levels <- TRUE
    mf[[1L]] <-
      quote(stats::model.frame) #mf[[1L]] <- as.name("model.frame") # make a model frame
    mf[[2L]] <-
      as.formula(paste(mf[[2L]][[2L]], "~.")) # keep all predictors
    YX <-
      eval(mf, parent.frame()) # data.frame containing response and ALL predictors

    # In this release not yet survival prediction
    if (method == "averaging") {
      # prediction-averaging method
      output <-
        .glm.mipred.cmb1.cv(formula, family, YX, nimp, folds, mice.options)
    } else {
      # Rubin's rules model
      output <-
        .glm.mipred.cmb2.cv(formula, family, YX, nimp, folds, mice.options)
    }

    row.names(output$pred) <-
      row.names(output$linpred) <- rownames.data
    output <- list(call = call,
      pred = output$pred,
      linpred = output$linpred)
    output
  }


R/mipred_cv.R
#' Cross-validation prediction using multiple imputation
#'
#' Calculates cross-validated predictions based on within-sample  assessment and calibration using
#' generalized linear models with multiple imputations to account for missing values
#' in predictor data.
#'
#' @param formula A formula object providing a symbolic description of the
#'   prediction model to be fitted.
#' @param family Specification of an appropriate error distribution and link
#'   function.
#' @param data A data.frame containing calibration data on \code{n} samples.
#'   Variables declared in \code{formula} must be found in \code{data}.
#' @param nimp Number of imputations used in the prediction of each observation.
#' @param folds Number of fold-partitions used in cross-validation. Put \code{folds=n}
#' for leave-one-out cross-validation.
#' @param method Imputation combination method. This defaults to
#'   \code{"averaging"} for the prediction-averaging approach. The alternative
#'   \code{"rubin"} applies the Rubin's rules pooled model.
#'
#' @return A list containing predictions.
#' \describe{\item{\code{pred}}{Matrix of predictions on the scale of the response
#' variable of dimension \code{n} by \code{nimp}.}
#' \item{\code{linpred}}{Matrix of predictions on the scale
#'   of the linear predictor of dimension \code{n} by \code{nimp}.}}
#'
#' @author Bart J A Mertens, \email{b.mertens@lumc.nl}
#' @references \url{https://arxiv.org/abs/1810.05099}
#' @seealso \code{\link{mice}}
#'
#' @examples
#' cll_bin<-cll # Generate a copy and construct binary outcome from survival information
#' cll_bin$srv5y_s[cll_bin$srv5y>12] <- 0  # Apply administrative censorship at t=12 months
#' cll_bin$srv5y[cll_bin$srv5y>12]  <- 12
#' cll_bin$Status[cll_bin$srv5y_s==1]<- 1  # Define the new binary "Status" outcome variable
#' cll_bin$Status[cll_bin$srv5y_s==0] <- 0  # As numeric -> 1:Dead, 0:Alive
#' cll_bin$Censor <- NULL # Remove survival outcomes
#' cll_bin$srv5y <- NULL
#' cll_bin$srv5y_s <- NULL
#'
#' # Cross-validate prediction using logistic regression in the first 100 samples
#' # Apply prediction-averaging using 5 imputations and 5 folds
#' output<-mipred.cv(Status~perfstat+remstat+cyto, family=binomial, data=cll_bin[1:100,],
#' nimp=5, folds=5)
#'
#' @export
#' @importFrom mice mice complete
#' @importFrom stats glm model.matrix
#'
mipred.cv <- function(formula, family, data, nimp, folds, method = "averaging"){

  call <- match.call()
  if (is.character(family))
    family <- get(family, mode = "function", envir = parent.frame())
  if (is.function(family))
    family <- family()
  if (is.null(family$family)) {
    print(family)
    stop("family' not recognized")
  }
  if (missing(data))
    data <- environment(formula)

  # generate calibration data.frame
  mf <- match.call(expand.dots = FALSE)
  m <- match(c("formula", "data", "subset", "weights"), names(mf), 0L)
  mf <- mf[c(1L, m)]
  mf$na.action=quote(na.pass)
  mf$drop.unused.levels <- TRUE
  mf[[1L]] <- quote(stats::model.frame) #mf[[1L]] <- as.name("model.frame")
  YX <- eval(mf, parent.frame()) # data.frame containing response and predictors

  # In this preliminary release only binary prediction
  if (method=="averaging"){ # prediction-averaging method
    output <- .glm_mipred_cmb1_cv(formula, family, YX, nimp, folds)
  } else { # Rubin's rules model
    output <- .glm_mipred_cmb2_cv(formula, family, YX, nimp, folds)
  }
  output<-list(call=call,pred=output$pred,linpred=output$linpred)
  output
}




R/mipred.R
#' Prediction using multiple imputation
#'
#' Calculates predictions from generalized linear models when multiple
#' imputations are used to account for missing values in predictor data.
#'
#' @param formula A formula object providing a symbolic description of the
#'   prediction model to be fitted.
#' @param family Specification of an appropriate error distribution and link
#'   function.
#' @param data A data.frame containing calibration data on \code{n} samples.
#'   Variables declared in \code{formula} must be found in \code{data}.
#' @param newdata A data.frame containing the predictors for observations to be
#'   predicted on \code{m} samples. This must have the same structure and
#'   variables as \code{data}, except for the outcome variable which is ignored
#'   in the construction of the predictions and can therefor be excluded from
#'   the object.
#' @param nimp Number of imputations used in the prediction of each observation.
#' @param folds Number of fold-partitions defined within \code{newdata}.
#' An integer from 1 to \code{m}. Defaults to NULL which internally
#' sets \code{folds=m}, which puts each observation in \code{newdata}
#' into its own singleton fold. The minimum value \code{folds=1}
#'  would predict the entire set \code{newdata} in a single step without partitioning.
#' @param method Imputation combination method. This defaults to
#'   \code{"averaging"} for the prediction-averaging approach. The alternative
#'   \code{"rubin"} applies the Rubin's rules pooled model.
#' @param mice.options Optional list containing arguments to be supplied to \code{mice}. Refer to the \code{mice} documentation for details.
#' The following options may be specified: \code{method}, \code{predictorMatrix}, \code{blocks},
#' \code{visitSequence}, \code{formulas}, \code{blots}, \code{post}, \code{defaultMethod},
#' \code{maxit}, \code{printFlag}, \code{seed}, \code{data.init}. Please refer to the
#' \code{mice} documentation for the description of these options. To set the number
#' of imputations \code{nimp} should be used. \code{seed} may be specified as a numeric vector
#' of length \code{nimp*folds} when \code{method} is set to \code{averaging} and of length \code{folds}
#' when \code{method} is set to \code{rubin}. Setting \code{seed} to a vector will cause each next
#' call to \code{mice} to use the next seed value in the vector. Setting the seed to a single
#' numeric value will cause all instances of
#' mice to use that same seed value. If you specify a seed vector of insufficient length
#' then the values will be recycled. The required length is \code{folds*nimp} for the averaging
#' approach and length \code{folds} for the rubin approach. The \code{defaultMethod} is set to
#' \code{c("pmm", "logreg", "polyreg", "polr")} by default. The default setting for
#' \code{printFlag} is FALSE. The default for \code{maxit} is 50. All other options are set
#' to \code{NULL} by default.
#'
#' @return A list containing predictions.
#' \describe{ \item{\code{pred}}{Matrix
#'   of predictions on the scale of the response variable of dimension \code{m}
#'   by \code{nimp}.} \item{\code{linpred}}{Matrix of predictions on the scale
#'   of the linear predictor of dimension \code{m} by \code{nimp}.} }
#'
#' @author Bart J A Mertens, \email{b.mertens@lumc.nl}
#' @references \url{https://arxiv.org/abs/1810.05099}
#' @seealso \code{\link{mice}}
#'
#' @examples
#' \donttest{
#' # Generate a copy of the cll data and construct binary outcome from survival information
#' cll_bin<-cll
#' cll_bin$srv5y_s[cll_bin$srv5y>12] <- 0  # Apply administrative censorship at t=12 months
#' cll_bin$srv5y[cll_bin$srv5y>12]  <- 12
#' cll_bin$Status[cll_bin$srv5y_s==1]<- 1  # Define the new binary "Status" outcome variable
#' cll_bin$Status[cll_bin$srv5y_s==0] <- 0  # As numeric -> 1:Dead, 0:Alive
#' cll_bin$Censor <- NULL # Remove survival outcomes
#' cll_bin$srv5y <- NULL
#' cll_bin$srv5y_s <- NULL
#'
#' # Predict observations 501 to 504 using the first 100 records to calibrate predictors
#' # Remove the identification variable before prediction calibration and imputation.
#' # Remove outcome for new observations
#' # Apply prediction-averaging using 5 imputations, set mice option maxit=5.
#' # Note these settings are only for illustration and should be set to higher values for
#' # practical use, particularly for nimp.
#' output<-mipred(Status ~ age10+cyto, family=binomial, data=cll_bin[1:100,-1],
#'   newdata=cll_bin[501:504,c(-1,-10)], nimp=5, mice.options=list(maxit=5))
#' }
#'
#' @export
#' @importFrom mice mice complete
#' @importFrom stats glm model.matrix
#'
mipred <-
  function(formula,
    family,
    data,
    newdata,
    nimp,
    folds = NULL,
    method = "averaging",
    mice.options = NULL) {
    call <- match.call()

    if (missing(formula))
      stop("formula not provided")
    if (!inherits(formula, "formula"))
      stop("formula argument not recognized as formula")

    if (is.character(family))
      family <-
        get(family, mode = "function", envir = parent.frame())
    if (is.function(family))
      family <- family()
    if (is.null(family$family)) {
      print(family)
      stop("family not recognized")
    }

    if (missing(data))
      stop("data not provided")
    if (!is.data.frame(data))
      stop("data should be a data frame", call. = FALSE)
    dup <- duplicated(colnames(data))
    if (any(dup))
      stop("Duplicate names found in data: ",
        paste(colnames(data)[dup], collapse = ", "))

    if (missing(newdata))
      stop("newdata not provided")
    if (!is.data.frame(newdata))
      stop("newdata should be a data frame", call. = FALSE)
    dupnew <- duplicated(colnames(newdata))
    if (any(dupnew))
      stop("Duplicate names found in newdata: ",
        paste(colnames(newdata)[dupnew], collapse = ", "))
    rownames.newdata <- rownames(newdata)

    if (missing(nimp))
      stop("nimp not provided")
    if (!is.numeric(nimp) | !is.vector(nimp)) {
      stop("nimp must be integer number greater than zero")
    } else {
      if (!(length(nimp) == 1)) {
        stop("nimp must be integer number greater than zero")
      } else{
        if ((as.integer(nimp) != nimp) | (nimp < 1))
          stop("nimp must be integer number greater than zero")
      }
    }

    if (is.null(folds)) {
      folds <-
        nrow(newdata) # force default to leave-one-out when NULL
    } else {
      if (!is.numeric(folds) | !is.vector(folds)) {
        stop("folds must be integer number greater than zero")
      } else {
        if (!(length(folds) == 1)) {
          stop("folds must be integer number greater than zero")
        } else{
          if ((as.integer(folds) != folds) | (folds < 1))
            stop("folds must be integer number greater than zero")
        }
      }
    }
    if (folds > nrow(newdata))
      stop("folds too large")

    if (is.null(mice.options)) {
      # setting default mice options
      mice.options[["method"]] <- NULL
      mice.options[["where"]] <- NULL
      mice.options[["visitSequence"]] <- NULL
      mice.options[["blots"]] <- NULL
      mice.options[["post"]] <- NULL
      mice.options[["defaultMethod"]] <-
        c("pmm", "logreg", "polyreg", "polr")
      mice.options[["maxit"]] <- 50
      mice.options[["printFlag"]] <- FALSE
      mice.options[["seed"]] <- NA
      mice.options[["data.init"]] <- NULL
    } else {
      # updating user provided mice options
      if (is.null(mice.options[["defaultMethod"]]))
        mice.options$defaultMethod <-
          c("pmm", "logreg", "polyreg", "polr")
      if (is.null(mice.options[["maxit"]]))
        mice.options$maxit <- 50
      if (is.null(mice.options[["printFlag"]]))
        mice.options$printFlag <- FALSE
      if (is.null(mice.options[["seed"]]))
        mice.options$seed <- NA
    }
    if (!is.null(mice.options[["m"]]))
      warning("mice.options argument m ignored, using nimp instead")

    if (all(!is.na(mice.options[["seed"]]))) {
      if (!is.vector(mice.options[["seed"]])) {
        stop("in mice.options seed must be a vector")
      } else {
        if (!is.numeric(mice.options[["seed"]])) {
          stop("in mice.options seed must be a numeric vector")
        } else {
          if (!all(as.integer(mice.options[["seed"]]) == mice.options[["seed"]]) |
              any(mice.options[["seed"]] < 0)) {
            stop("in mice.options seed must be positive integer")
          } else{
            if (method == "rubin") {
              mice.options[["seed"]] <-
                matrix(mice.options[["seed"]],
                  nrow = folds,
                  ncol = 1)[, 1]
            } else {
              mice.options[["seed"]] <-
                matrix(mice.options[["seed"]],
                  nrow = folds * nimp,
                  ncol = 1)[, 1]
            }
          }
        }
      }
    }

    # generate calibration data.frame
    mf <- match.call(expand.dots = FALSE)
    m <-
      match(c("formula", "data", "subset", "weights"), names(mf), 0L)
    mf <- mf[c(1L, m)]
    mf$na.action = quote(na.pass)
    mf$drop.unused.levels <- TRUE
    mf[[1L]] <-
      quote(stats::model.frame) #mf[[1L]] <- as.name("model.frame") # make a model frame
    mf[[2L]] <-
      as.formula(paste(mf[[2L]][[2L]], "~.")) # keep all predictors
    YX <-
      eval(mf, parent.frame()) # data.frame containing response and ALL predictors

    # generate validation data.frame
    mf <- match.call(expand.dots = FALSE)
    m <- match(c("formula", "newdata"), names(mf), 0L)
    mf <- mf[c(1L, m)]
    mf$na.action = quote(na.pass)
    mf$drop.unused.levels <- TRUE
    names(mf)[3L] <-
      "data" # change to "data" to setup call to model.frame for predictor data
    mf[[2L]] <- as.formula("~.") # keep all predictors
    mf[[1L]] <-
      quote(stats::model.frame) #mf[[1L]] <- as.name("model.frame")
    Xnew <-
      eval(mf, parent.frame()) # data.frame containing ALL predictors only

    YX.vars <- attributes(attributes(YX)$terms)$term.labels
    Xnew.vars <- attributes(attributes(Xnew)$terms)$term.labels

    if (!setequal(YX.vars, Xnew.vars))
      #check for predictors compatibility between calibration and validation data
      stop("data and newdata not compatible: check consistency of included predictors")

    # In this release not yet survival prediction
    if (method == "averaging") {
      # prediction-averaging method
      output <-
        .glm.mipred.cmb1(formula, family, YX, Xnew, nimp, folds, mice.options)
    } else {
      # Rubin's rules model
      output <-
        .glm.mipred.cmb2(formula, family, YX, Xnew, nimp, folds, mice.options)
    }

    row.names(output$pred) <-
      row.names(output$linpred) <- rownames.newdata
    output <- list(call = call,
      pred = output$pred,
      linpred = output$linpred)
    output
  }
